plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,241 @@
+
+	var err error
+	if f.tlsConf != nil {
+		f.listener, err = tls.Listen("tcp", f.ServiceAddress, f.tlsConf)
+	} else {
+		f.listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+
+	f.server = http.Server{
+		Addr:         f.ServiceAddress,
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+
+	go func() {
+		if err := f.server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("starting server failed: %v", err)
+			}
+		}
+	}()
+
+	f.Log.Infof("listening on %s", f.listener.Addr().String())
+
+	return nil
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	err := f.server.Shutdown(context.Background())
+	if err != nil {
+		f.Log.Infof("shutting down server failed: %v", err)
+	}
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	if !slices.Contains(f.Paths, req.URL.Path) {
+		res.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	requestID := req.Header.Get("x-amz-firehose-request-id")
+	r := &request{req: req}
+
+	if err := r.validate(); err != nil {
+		f.Log.Errorf("validation failed: %v", err)
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+		}
+		return
+	}
+
+	if err := r.authenticate(f.AccessKey); err != nil {
+		f.Log.Errorf("authentication failed: %v", err)
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+		}
+		return
+	}
+
+	data, err := r.decodeData()
+	if err != nil {
+		f.Log.Errorf("decode base64 data from request %q failed: %v", requestID, err)
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+		}
+		return
+	}
+
+	var metrics []telegraf.Metric
+	for _, bytes := range data {
+		m, err := f.parser.Parse(bytes)
+		if err != nil {
+			f.Log.Errorf("parse data from request %q failed: %v", requestID, err)
+			// respond with bad request status code to inform firehose about the failure
+			r.res.statusCode = http.StatusBadRequest
+			if err = r.sendResponse(res); err != nil {
+				f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+			}
+			return
+		}
+		metrics = append(metrics, m...)
+	}
+
+	if len(metrics) == 0 {
+		f.once.Do(func() {
+			f.Log.Info(internal.NoMetricsCreatedMsg)
+		})
+		return
+	}
+
+	if err := f.addParameterTags(r, metrics); err != nil {
+		f.Log.Errorf("add parameter tags for request %q failed: %v", requestID, err)
+		r.res.statusCode = http.StatusBadRequest
+		if err = r.sendResponse(res); err != nil {
+			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
+		}
+		return
+	}
+
+	for _, m := range metrics {
+		m.AddTag("firehose_http_path", req.URL.Path)
+	}
+
+	for _, m := range metrics {
+		f.acc.AddMetric(m)
+	}
```

Comment:
How about
```suggestion
	tags, err := r.extractParameterTags(f.ParameterTags)
	if err != nil {
		f.Log.Errorf("extracting parameter tags for request %q failed: %v", requestID, err)
		r.res.statusCode = http.StatusBadRequest
		if err = r.sendResponse(res); err != nil {
			f.Log.Errorf("sending response to request %q failed: %v", requestID, err)
		}
		return
	}

	for _, m := range metrics {
		for k, v := range tags {
			m.AddTag(k, v)
		}
		m.AddTag("firehose_http_path", req.URL.Path)
		f.acc.AddMetric(m)
	}
```

This way you don't need to iterate over the metrics three times and do have a clear code structure where everything related to processing the request is in the request "class"...

Reply:
makes sense